src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -192,13 +192,14 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
                     {
                         var entry = item.Entry;
                         var matcher = item.TemplateMatcher;
-                        if (!matcher.TryMatch(context.HttpContext.Request.Path, context.RouteData.Values))
-                        {
-                            continue;
-                        }
 
                         try
                         {
+                            if (!matcher.TryMatch(context.HttpContext.Request.Path, context.RouteData.Values))
```

Comment:
Yeah - the question of whether or not to do cleanup on exception has multiple answers. I decided to just go ahead and clean up when it throws even though no one will look at the data.

I felt like coding it this way asks fewer questions than the other way, as people are used to seeing the try/finally question.